### PR TITLE
[AIRFLOW-1985] Changes to propagate config in subdags

### PR DESCRIPTION
**Pull request by edgarRd**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1985

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Propagating configuration within subdags when using impersonation to
guarantee correct behavior of tasks.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Added relevant tests.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

@aoen PTAL

### CODE REVIEW

File: airflow/operators/subdag_operator.py [hunk 1]
```diff
@@ -84,9 +86,17 @@ def __init__(
...
 
     def execute(self, context):
         ed = context['execution_date']
         self.subdag.run(
             start_date=ed, end_date=ed, donot_pickle=True,
-            executor=self.executor)
+            executor=self.executor, cfg_path=self.cfg_path)
+
+    def post_execute(self, context, result=None):
+        if self.run_as_user and self.cfg_path:
```
Comment (aoen) [inline airflow/operators/subdag_operator.py]: Any reason not to put this in execute itself? That way you can open the file using a "with" instead which is a bit safer in terms of automatic cleanup.

File: airflow/executors/base_executor.py [hunk 2]
```diff
@@ -49,16 +49,16 @@ def queue_command(self, task_instance, command, priority=1, queue=None):
...
-            self,
-            task_instance,
-            mark_success=False,
-            pickle_id=None,
-            ignore_all_deps=False,
-            ignore_depends_on_past=False,
-            ignore_task_deps=False,
-            ignore_ti_state=False,
-            pool=None):
+    def queue_task_instance(self,
```
Comment (aoen) [inline airflow/executors/base_executor.py:52]: Let's add an explanation why cfg_path is needed and a JIRA for the appropriate fix as discussed.

File: airflow/operators/subdag_operator.py [hunk 3]
```diff
@@ -84,9 +86,17 @@ def __init__(
 
         self.subdag = subdag
         self.executor = executor
+        self.cfg_path = None  # will be populated if running with impersonation
+
+    def pre_execute(self, context):
+        self.cfg_path = configuration_subset(MAIN_SECTIONS) if self.run_as_user else None
```
Comment (aoen) [inline airflow/operators/subdag_operator.py]: Why copy a subset instead of the whole thing?

File: airflow/executors/base_executor.py (hunk 2 above)
Comment (edgarRd) [reply airflow/executors/base_executor.py:52]: Added.

**Comment by edgarRd** [comment]
Addressed comments, @aoen PTAL

**Comment by aoen** [comment]
@edgarRd there was one more comment, Why copy a subset of the config instead of the whole thing (trying to reduce the complexity of this change since it's a work-around)?

**Comment by edgarRd** [comment]
I tried initially copying the whole configuration settings but it's not possible given that sometimes configuration elements come from env variables, such as `testsection` when running tests, and these can't be serialized since they could be arbitrary. At the end, you still need some sort of logic to rule these parts out.

I did reduce the complexity of this change by always copying the configuration sections when running within the context of the subdag operator (using a BackfillJob) - the generated configuration files are always removed by the `airflow run` in [cli.py#L343](https://github.com/apache/incubator-airflow/blob/master/airflow/bin/cli.py#L343).

**Comment by edgarRd** [comment]
This changes were already merged in https://github.com/apache/incubator-airflow/pull/2991. Closing.

**Pull request closed without merging**